backend/runner/query/client.go

```diff
@@ -0,0 +1,312 @@
+	client, err := newClient(dbName)
+	if err != nil {
+		return err
+	}
+	_, err = client.performQuery(ctx, nil, querypb.CommandType_COMMAND_TYPE_EXEC, rawSQL, params, colFieldNames)
+	return err
+}
+
+type client struct {
+	queryconnect.QueryServiceClient
+}
+
+func newClient(dbName string) (*client, error) {
+	address := os.Getenv(strings.ToUpper("FTL_QUERY_ADDRESS_" + dbName))
+	if address == "" {
+		return nil, fmt.Errorf("query address for %s not found", dbName)
+	}
+	return &client{QueryServiceClient: rpc.Dial(queryconnect.NewQueryServiceClient, address, log.Error)}, nil
+}
+
+// performQuery performs a SQL query and returns the results.
+//
+// note: accepts colFieldNames as []tuple.Pair[string, string] rather than map[string]string to preserve column order
+func (c *client) performQuery(ctx context.Context, respType reflect.Type, commandType querypb.CommandType, rawSQL string, params []any, colFieldNames []tuple.Pair[string, string]) ([]any, error) {
+	sqlParams := make([]*querypb.SQLValue, 0, len(params))
+	for _, param := range params {
+		sqlValue, err := convertParamToSQLValue(param)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert parameter %v of type %T: %w", param, param, err)
+		}
+		sqlParams = append(sqlParams, sqlValue)
+	}
+
+	colToFieldName := map[string]string{}
+	for _, pair := range colFieldNames {
+		colToFieldName[pair.A] = pair.B
+	}
+
+	resultCols := make([]string, 0, len(colFieldNames))
+	for _, pair := range colFieldNames {
+		resultCols = append(resultCols, pair.A)
+	}
+
+	stream, err := c.QueryServiceClient.ExecuteQuery(ctx, connect.NewRequest(&querypb.ExecuteQueryRequest{
+		RawSql:        rawSQL,
+		CommandType:   commandType,
+		Parameters:    sqlParams,
+		ResultColumns: resultCols,
+	}))
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute query: %w", err)
+	}
+
+	var results []any
+	for stream.Receive() {
+		resp := stream.Msg()
+		switch r := resp.Result.(type) {
+		case *querypb.ExecuteQueryResponse_ExecResult:
+			return nil, nil
+		case *querypb.ExecuteQueryResponse_RowResults:
+			if len(r.RowResults.Rows) == 0 {
+				continue
+			}
+
+			respValue := reflect.New(respType).Elem()
+			for col, field := range colToFieldName {
+				if sqlVal, ok := r.RowResults.Rows[col]; ok {
+					fieldValue := respValue.FieldByName(field)
+					if !fieldValue.IsValid() {
+						continue
+					}
+					if err := setFieldFromSQLValue(fieldValue, sqlVal); err != nil {
+						return nil, fmt.Errorf("failed to set field %s from column %s: value type %T = %v: %w",
+							field, col, sqlVal.Value, sqlVal.Value, err)
+					}
+				}
+			}
+			results = append(results, respValue.Interface())
+		}
+	}
+	if err := stream.Err(); err != nil {
+		return nil, fmt.Errorf("failed to receive query results: %w", err)
+	}
+	return results, nil
+}
+
+func setFieldFromSQLValue(field reflect.Value, sqlValue *querypb.SQLValue) error {
```

Comment:
Related to my question above, if we use the SQL drivers directly we don't need any of this code, as we can scan directly into fields.